### PR TITLE
fix: support callv and proxies for lsp.js

### PR DESCRIPTION
**Pull request by ajbt200128**
Previously we added support for proxies in #9994. But it broke LSP.js networking as we were forced to use callv to support proxies as a workaround, and callv isn't supported in cohttp-jsoo!. So we have to patch callv and patch our XHR js library to enable proxies for this stuff. I'm going to try and make a bunch of PRs to cohttp and ca-certs etc to fix this upstream, but for now we have this fix.

### For testing js w/out proxy
```bash
make build-semgrep-jsoo-debug
make -C js build
make -C js test
```
### For testing JS + proxy
```bash
make build-semgrep-jsoo-debug
make -C js build
make -C js/language_server
# copy js/language_server/dist into extension
# Run vscode with proxy and Use JS to confirm it works
```

TODO add MIT license to VSCode

### CODE REVIEW

File: js/node_shared/XMLHttpRequest.js [hunk 1]
```diff
@@ -385,6 +385,17 @@ XMLHttpRequest = function () {
       headers["Content-Length"] = 0;
     }
 
+    // get if HTTP_PROXY or HTTPS_PROXY is set
+    var proxy = process.env[ssl ? "HTTPS_PROXY" : "HTTP_PROXY"];
```
Comment (kopecs) [inline js/node_shared/XMLHttpRequest.js]:
```suggestion
    const proxy = process.env[ssl ? "HTTPS_PROXY" : "HTTP_PROXY"];
```

File: js/node_shared/Semgrep_js_node_shared.ml [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+(*****************************************************************************)
+
+module Patched_cohttp_lwt_jsoo : Cohttp_lwt.S.Client = struct
+  include Cohttp_lwt_jsoo.Client
+
+  (* Callv isn't currently supported on the JS side so we have to provide it ourselves :( *)
+  let callv ?ctx (uri : Uri.t) stream =
+    (* Differs from request uri if HTTP(S)_PROXY is set. But we handle this in
+       the node_shared XHR*)
+    ignore uri;
```
Comment (kopecs) [inline js/node_shared/Semgrep_js_node_shared.ml]:
```suggestion
```
Dead now?

File: js/node_shared/XMLHttpRequest.js [hunk 3]
```diff
@@ -13,6 +13,26 @@
  * @author Dan DeFelippi <dan@driverdan.com>
  * @contributor David Ellis <d.f.ellis@ieee.org>
  * @license MIT
```
Comment (kopecs) [inline js/node_shared/XMLHttpRequest.js:18]: Is this whole file copied, or is it modified a fair amount? I would recommend having a note so we know what's what (like precisely what is just inbound versus derived versus ours and unrelated -- best of the last isn't in the same file but might make sense)

File: src/osemgrep/language_server/Test_LS_e2e.ml [hunk 4]
```diff
@@ -664,6 +669,9 @@ let check_startup info folders (files : Fpath.t list) =
   let%lwt () = assert_progress info "Refreshing Rules" in
   let%lwt rulesRefreshedNotif = receive_notification info in
   let%lwt () = assert_notif rulesRefreshedNotif "semgrep/rulesRefreshed" in
+  let rules_length = List.length info.server.session.cached_session.rules in
+  Alcotest.(check bool)
+    "rules refreshed greater than 1" (rules_length > 20) true;
```
Comment (kopecs) [inline src/osemgrep/language_server/Test_LS_e2e.ml]: This is a test so it's probably fine but fyi we have a function to check if a list is shorter/longer than a fixed length if we don't need to count every element.

File: js/language_server/dune [hunk 5]
```diff
@@ -19,13 +19,11 @@
   lwt_platform.js
 
    networking.http_helpers
-   cohttp-lwt-jsoo
   )
  (modes js)
  (js_of_ocaml
   (javascript_files
    ;; Custom XMLHttpRequest implementation that doesn't preset the User-Agent header
```
Comment (kopecs) [inline js/language_server/dune]:
```suggestion
```
Should also be removed given the deletion on the next line?

File: js/node_shared/Semgrep_js_node_shared.ml [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+(* Commentary *)
+(*  *)
```
Comment (kopecs) [inline js/node_shared/Semgrep_js_node_shared.ml]: A longer comment here explaining the use of `callv` (and anything else we later add) would be nice.

**Review by kopecs — APPROVED** [review]
Seems mostly good. I think we would benefit from some more documentation.

**Comment by ajbt200128** [comment]
pro checks failing seem unrelated

**Pull request merged**